slither/utils/upgradeability.py

```diff
@@ -0,0 +1,570 @@
+    ]
+    if len(order_vars2) < len(order_vars1):
+        for variable in order_vars1:
+            if variable.name not in [v.name for v in order_vars2]:
+                results.append(variable)
+    return results
+
+
+def is_function_modified(f1: Function, f2: Function) -> bool:
+    """
+    Compares two versions of a function, and returns True if the function has been modified.
+    First checks whether the functions' content hashes are equal to quickly rule out identical functions.
+    Walks the CFGs and compares IR operations if hashes differ to rule out false positives, i.e., from changed comments.
+
+    Args:
+        f1: Original version of the function
+        f2: New version of the function
+
+    Returns:
+        True if the functions differ, otherwise False
+    """
+    # If the function content hashes are the same, no need to investigate the function further
+    if f1.source_mapping.content_hash == f2.source_mapping.content_hash:
+        return False
+    # If the hashes differ, it is possible a change in a name or in a comment could be the only difference
+    # So we need to resort to walking through the CFG and comparing the IR operations
+    queue_f1 = [f1.entry_point]
+    queue_f2 = [f2.entry_point]
+    visited = []
+    while len(queue_f1) > 0 and len(queue_f2) > 0:
+        node_f1 = queue_f1.pop(0)
+        node_f2 = queue_f2.pop(0)
+        visited.extend([node_f1, node_f2])
+        queue_f1.extend(son for son in node_f1.sons if son not in visited)
+        queue_f2.extend(son for son in node_f2.sons if son not in visited)
+        for i, ir in enumerate(node_f1.irs):
+            if encode_ir_for_compare(ir) != encode_ir_for_compare(node_f2.irs[i]):
+                return True
+    return False
+
+
+# pylint: disable=too-many-branches
+def ntype(_type: Union[Type, str]) -> str:
+    if isinstance(_type, ElementaryType):
+        _type = str(_type)
+    elif isinstance(_type, ArrayType):
+        if isinstance(_type.type, ElementaryType):
+            _type = str(_type)
+        else:
+            _type = "user_defined_array"
+    elif isinstance(_type, Structure):
+        _type = str(_type)
+    elif isinstance(_type, Enum):
+        _type = str(_type)
+    elif isinstance(_type, MappingType):
+        _type = str(_type)
+    elif isinstance(_type, UserDefinedType):
+        _type = "user_defined_type"  # TODO: this could be Contract, Enum or Struct
+    else:
+        _type = str(_type)
+
+    _type = _type.replace(" memory", "")
+    _type = _type.replace(" storage ref", "")
+
+    if "struct" in _type:
+        return "struct"
+    if "enum" in _type:
+        return "enum"
+    if "tuple" in _type:
+        return "tuple"
+    if "contract" in _type:
+        return "contract"
+    if "mapping" in _type:
+        return "mapping"
+    return _type.replace(" ", "_")
+
+
+# pylint: disable=too-many-branches
+def encode_ir_for_compare(ir: Union[Operation, Variable]) -> str:
```

Comment:
It's great that we have added the types, because this does not sound right :) 

Should we split this function in two?
- `encore_ir_for_compare`
- `encore_variable_for_compare`

Note that we probably have the same issue in the slither-similar code, but that should be fixed in another PR

Reply:
>Should we split this function in two?

Done!